### PR TITLE
Push some deferred wrangling down into DeferredCache

### PR DESCRIPTION
**Pull request by richvdh**
Rather than `DeferredCache.get()` sometimes returning an `ObservableDeferred` that you have to `.observe`, and sometimes returning an immediate result that you have to wrap with a `Deferred`, we can make it behave consistently. This removes complexity from the call site, improves type safety, and generally leasds to a cleaner API.

~~Based on #8568~~

### CODE REVIEW

File: synapse/util/caches/descriptors.py [hunk 1]
```diff
@@ -208,26 +207,12 @@ def _wrapped(*args, **kwargs):
...
+                ret = cache.get(cache_key, callback=invalidate_callback)
             except KeyError:
                 ret = defer.maybeDeferred(preserve_fn(self.orig), obj, *args, **kwargs)
+                cache.set(cache_key, ret, callback=invalidate_callback)
 
-                def onErr(f):
-                    cache.invalidate(cache_key)
-                    return f
-
-                ret.addErrback(onErr)
```
Comment (richvdh) [inline synapse/util/caches/descriptors.py:225]: this happens anyway, at https://github.com/matrix-org/synapse/blob/e67f5720944687ebed524408c08076b35178f2b3/synapse/util/caches/deferred_cache.py#L264.

Comment (richvdh) [reply synapse/util/caches/descriptors.py:225]: ohh, except that's slightly different. I better write some tests.

File: synapse/util/caches/deferred_cache.py [hunk 2]
```diff
@@ -119,21 +128,27 @@ def check_thread(self):
     def get(
         self,
         key: KT,
-        default=_Sentinel.sentinel,
         callback: Optional[Callable[[], None]] = None,
         update_metrics: bool = True,
-    ):
+    ) -> defer.Deferred:
```
Comment (clokep) [inline synapse/util/caches/deferred_cache.py:133]: I had debated making this async at some point, which should work if all callers just await it, which they might now that you added `get_immediate`?

Comment (richvdh) [reply synapse/util/caches/deferred_cache.py:133]: well, the only place this is called is https://github.com/matrix-org/synapse/pull/8572/files#diff-3aa4bc9d8830ffd035d53b8af86b548a21f2a4cddb3e5240445661614a88a08fR313, which doesn't await it. In any case the logcontext-wrangling means you end up needing to wrap it with `make_deferred_yieldable`.

I considered making `get()` do the `make_deferred_yieldable` call itself, but decided that the resultant asymmetry was confusing.

File: synapse/util/caches/descriptors.py (hunk 1 above)
Comment (richvdh) [reply synapse/util/caches/descriptors.py:225]: ok, I've stared at this harder and am pretty sure it's redundant. As noted above, if the function fails, then we call `entry.invalidate()` on the "temporary" entry in the pending_deferred_cache. The only difference between that and calling `cache.invalidate(cache_key)` as shown here would be if there was also an entry in the *main* cache. But we know that can't be the case, because we checked it at line 210, and there is no way for it to have been populated in the meantime.

https://github.com/matrix-org/synapse/pull/8572/files#diff-3aa4bc9d8830ffd035d53b8af86b548a21f2a4cddb3e5240445661614a88a08fR210

(unless someone decided to sidestep the CacheDescriptor and manually call `cache.set` or `cache.prefill` while this request was running? but that would be silly, and even if they did it's not obvious that we should invalidate it here.)

In short, I think this is handled correctly in DeferredCache itself.

File: tests/util/caches/test_descriptors.py [hunk 3]
```diff
@@ -470,9 +470,9 @@ def func(self, key):
 
         a = A()
 
-        a.func.prefill(("foo",), ObservableDeferred(d))
```
Comment (richvdh) [inline tests/util/caches/test_descriptors.py]: I don't think that prefilling with an ObservableDeferred is the way this API is meant to be used. At least, it hasn't been used that way for a long time.

**Comment by richvdh** [comment]
There's a fair bit going on in this PR, but the diff looks worse than it is thanks to https://github.com/matrix-org/synapse/pull/8572/commits/438ffc494420c8afabfbeb991d1916ad678ad7b3 which is a lift & shift of an entire test class, so hopefully it is reasonably easy to review commit-by-commit.

File: synapse/util/caches/deferred_cache.py [hunk 4]
```diff
@@ -130,16 +130,22 @@ def get(
...
 
+        For symmetry with set(), this method does *not* follow the synapse logcontext
+        rules: the logcontext will not be cleared on return, and the Deferred will run
+        its callbacks in the sentinel context. In other words: wrap the result with
+        make_deferred_yieldable() before `await`ing it.
+
         Args:
-            key(tuple)
+            key:
             callback(fn): Gets called when the entry in the cache is invalidated
```
Comment (clokep) [inline synapse/util/caches/deferred_cache.py:143]:
```suggestion
            callback: Gets called when the entry in the cache is invalidated
```

File: synapse/util/caches/descriptors.py [hunk 5]
```diff
@@ -286,7 +271,7 @@ def __init__(self, orig, cached_method_name, list_name, num_args=None):
 
     def __get__(self, obj, objtype=None):
         cached_method = getattr(obj, self.cached_method_name)
-        cache = cached_method.cache
+        cache = cached_method.cache  # type: DeferredCache[Tuple, Any]
```
Comment (clokep) [inline synapse/util/caches/descriptors.py:274]: Should this also be `CacheKey`, not `Tuple`?

```suggestion
        cache = cached_method.cache  # type: DeferredCache[CacheKey, Any]
```

File: tests/util/caches/test_deferred_cache.py [hunk 6]
```diff
@@ -39,6 +39,99 @@ def test_hit(self):
...
+        self.assertSetEqual(callbacks, set())
+
+        # now fire off all the deferreds
+        origin_d.callback(20)
+        self.assertEqual(self.successResultOf(set_d), 20)
+        self.assertEqual(self.successResultOf(get_d), 20)
+
+        # now the original invalidation callback should have been called, but none of
+        # the others
+        self.assertSetEqual(callbacks, {"prefill"})
```
Comment (clokep) [inline tests/util/caches/test_deferred_cache.py:93]: Curious why you used `assertSetEqual` here? Was `assertEqual` not working for some reason / does this have slightly different behavior I'm unaware of?

File: tests/util/caches/test_descriptors.py [hunk 7]
```diff
@@ -130,6 +131,56 @@ def fn(self, arg1):
...
+        # ... and the callbacks to have been, uh, called.
+        self.assertSetEqual(callbacks, {"d1", "d2"})
+
+        # ... leaving the cache empty
+        self.assertEqual(len(obj.fn.cache.cache), 0)
+
+        # and a second call should work as normal
+        obj.result = defer.succeed(100)
+        d3 = obj.fn(1)
+        self.assertEqual(self.successResultOf(d3), 100)
```
Comment (clokep) [inline tests/util/caches/test_descriptors.py:182]: I believe that this now calls into the function again, should we assert `obj.call_count == 2`?

File: tests/util/caches/test_deferred_cache.py [hunk 8]
```diff
@@ -53,7 +53,9 @@ def check1(r):
             self.assertTrue(set_d.called)
             return r
 
-        get_d.addCallback(check1)
+        # TODO: Actually ObservableDeferred *doesn't* run its tests in order on py3.8.
+        #   maybe we should fix that?
+        # get_d.addCallback(check1)
```
Comment (clokep) [inline tests/util/caches/test_deferred_cache.py:58]: Yeah this is confusing. I'm not sure it is worth fixing (in this PR, at least). Do you see any downsides to this in practice?

**Review by clokep — APPROVED** [review]

File: synapse/util/caches/descriptors.py (hunk 5 above)
Comment (richvdh) [reply synapse/util/caches/descriptors.py:274]: yes, thanks.

File: tests/util/caches/test_deferred_cache.py (hunk 8 above)
Comment (richvdh) [reply tests/util/caches/test_deferred_cache.py:58]: well, at some level I suspect it means we end up processing requests that arrived later before those that arrived earlier, which isn't exactly forbidden but does feel like it might lead to some surprising behaviour.

But likewise, I'm not going to rush into fixing it right now.

File: tests/util/caches/test_deferred_cache.py (hunk 6 above)
Comment (richvdh) [reply tests/util/caches/test_deferred_cache.py:93]: I wanted to make sure that they were compared as sets, rather than (say) lists which might cause false differences due to ordering.  But I didn't think about it hard enough... I'll switch to `assertEqual` for consistency.

File: tests/util/caches/test_descriptors.py (hunk 7 above)
Comment (richvdh) [reply tests/util/caches/test_descriptors.py:182]: sure, why not

**Comment by richvdh** [comment]
I'm going to rebase this into useful commits before merging.
